File: backend/app/schemas/headlines.py
"""Contract for the homepage ticker's candidate pool."""

from datetime import date as date_type
from typing import Literal

from pydantic import BaseModel


class HeadlineToken(BaseModel):
    """A span of `text` naming an entity, for the client to tint.

    Carries the code, never a colour: one palette, resolved from standings, has
    to govern the whole page.
    """

    start: int
    end: int
    kind: Literal["driver", "team"]
    code: str


class HeadlineValidity(BaseModel):
    """When a candidate stops being true enough to show.

    The pool is rebuilt from results rather than patched, so this tells a
    cached client how long the copy it holds can be trusted.
    """

    kind: Literal["next_race", "season_end", "date"]
    date: date_type | None = None


class Headline(BaseModel):
    """One claim about Formula 1 that the database can prove."""

    id: str
    category: str
    kicker: str
    text: str
    tokens: list[HeadlineToken] = []
    weight: float
    valid_until: HeadlineValidity
    href: str | None = None


class HeadlinesResponse(BaseModel):
    """Response for GET /api/headlines.

    The whole candidate pool. Selection, ordering and the daily shuffle run on
    the client; this endpoint only decides what is true.
    """

    season: int
    round: int | None
    headlines: list[Headline]

File: frontend/lib/queries/headlines.ts
import { queryOptions } from "@tanstack/react-query";
import { minutes } from "./durations";
import { getJson } from "./http";

/** A span of `text` naming an entity, for the client to tint. */
export type HeadlineToken = {
  start: number;
  end: number;
  kind: "driver" | "team";
  code: string;
};

export type HeadlineValidity = {
  kind: "next_race" | "season_end" | "date";
  date: string | null;
};

export type Headline = {
  id: string;
  category: string;
  kicker: string;
  text: string;
  tokens: HeadlineToken[];
  weight: number;
  valid_until: HeadlineValidity;
  href: string | null;
};

export type HeadlinesResponse = {
  season: number;
  round: number | null;
  headlines: Headline[];
};

export const headlineKeys = {
  season: (season: number) => ["headlines", season] as const,
};

export function headlinesQuery(season: number | null) {
  return queryOptions({
    queryKey: headlineKeys.season(season ?? 0),
    queryFn: () =>
      getJson<HeadlinesResponse>(
        `/api/headlines?season=${season}`,
        "Failed to fetch headlines",
      ),
    enabled: season !== null,
    staleTime: minutes(10),
  });
}
